craft_parts/executor/part_handler.py

```diff
@@ -245,6 +244,131 @@ def _run_step(
 
         return step_handler.run_builtin()
 
+    def _update_action(self, action: Action, *, step_info: StepInfo) -> None:
+        handler: Callable[[StepInfo], None]
+
+        if action.step == Step.PULL:
+            handler = self._update_pull
+        elif action.step == Step.BUILD:
+            handler = self._update_build
+        else:
+            step_name = action.step.name.lower()
+            raise errors.InvalidAction(
+                f"cannot update step {step_name!r} of {self._part.name!r}."
+            )
+
+        callbacks.run_pre_step(step_info)
+        handler(step_info)
+        state_file = states.state_file_path(self._part, action.step)
+        state_file.touch()
+        callbacks.run_post_step(step_info)
+
+    def _update_pull(self, step_info: StepInfo) -> None:
+        self._make_dirs()
+
+        # if there's an override-pull scriptlet, execute it instead
+        if self._part.spec.override_pull:
+            self._run_step(
+                step_info=step_info,
+                scriptlet_name="override-pull",
+                work_dir=self._part.part_src_dir,
+            )
+            return
+
+        # consistency check
+        if not self._source_handler:
```

Comment:
The sequencer won't generate inconsistent actions, but test to prevent problems caused by actions added by hand. This also keeps linters happy.

Reply:
That's because the spec for `get_source_handler` returns Optional source handler.  (that is, `source` was not specified`)

Reply:
Is that a problem for this?  Or does _update_pull() never get invoked if no source?

Reply:
Update events are not generated by craft-parts for parts without `source` (for example when using the `nil` plugin), but it's possible to create `Action("part_name", Step.PULL, ActionType.UPDATE)` for that part and send it to the action executor. In this case it will raise an exception (it's `RuntimeError` in the code, but I think it should be `InvalidAction`, will update accordingly).

Reply:
Or perhaps we should just log the attempt and do nothing.

Reply:
What if there was a source for a part, then removed?

Reply:
The part specification is re-evaluated at each run, so if source was removed it will be considered non-existent.